arangod/IResearch/IResearchDataStore.cpp

```diff
@@ -1500,109 +1505,55 @@ void IResearchDataStore::properties(LinkLock linkLock,
-
-    auto ptr = std::make_unique<IResearchTrxState>(
-        std::move(linkLock), *(_dataStore._writer), nested);
-
-    ctx = ptr.get();
-    state.cookie(key, std::move(ptr));
-    if (!ctx) {
-      return {
-          TRI_ERROR_INTERNAL,
-          absl::StrCat(
-              "failed to store state into a TransactionState for remove from "
-              "ArangoSearch index '",
-              index().id().id(), "', tid '", state.id().id(), "', documentId '",
-              documentId.id(), "'")};
-    }
-    state.addBeforeCommitCallback(&_beforeCommitCallback);
-    state.addAfterCommitCallback(&_afterCommitCallback);
-  }
-
-  // ...........................................................................
-  // if an exception occurs below than the transaction is dropped including
-  // all all of its fid stores, no impact to iResearch View data integrity
-  // ...........................................................................
-  try {
-    ctx->remove(documentId);
-
-    return {TRI_ERROR_NO_ERROR};
-  } catch (basics::Exception const& e) {
-    return {e.code(), absl::StrCat("caught exception while removing document "
-                                   "from ArangoSearch index '",
-                                   index().id().id(), "', documentId '",
-                                   documentId.id(), "': ", e.message())};
-  } catch (std::exception const& e) {
-    return {TRI_ERROR_INTERNAL,
-            absl::StrCat("caught exception while removing document from "
-                         "ArangoSearch index '",
-                         index().id().id(), "', documentId '", documentId.id(),
-                         "': ", e.what())};
-  } catch (...) {
-    return {TRI_ERROR_INTERNAL,
-            absl::StrCat("caught exception while removing document from "
-                         "ArangoSearch index '",
-                         index().id().id(), "', documentId '", documentId.id(),
-                         "'")};
+  auto* ctx = getContext(state);
+  if (ADB_UNLIKELY(!ctx)) {
+    return {TRI_ERROR_ARANGO_INDEX_HANDLE_BAD,
+            MakeMessage("While removing document failed to lock", index(),
+                        state, documentId)};
   }
-
+  ctx->remove(documentId);
   return {};
 }
 
-bool IResearchDataStore::exists(IResearchDataStore::Snapshot const& snapshot,
-                                LocalDocumentId documentId,
-                                TRI_voc_tick_t const* recoveryTick) const {
-  if (recoveryTick == nullptr) {
-    // skip recovery check
-  } else if (*recoveryTick <= _dataStore._recoveryTickLow) {
-    LOG_TOPIC("6e128", TRACE, TOPIC)
-        << "skipping 'exists', operation tick '" << *recoveryTick
-        << "', recovery tick low '" << _dataStore._recoveryTickLow << "'";
+void IResearchDataStore::recoveryRemove(LocalDocumentId documentId) {
+  TRI_ASSERT(!isOutOfSync());
+  TRI_ASSERT(_recoveryRemoves);
+  _recoveryRemoves->emplace(documentId);
```

Comment:
This buffers all removes for the recovery in main memory.
What if there are a lot of remove operations to recover? Wouldn't this use a large amount of main memory then, and could lead to OOM?

Reply:
> This buffers all removes for the recovery in main memory.
What if there are a lot of remove operations to recover? Wouldn't this use a large amount of main memory then, and could lead to OOM?

Yes it's possible, but it was possible (but was easy in the 10~ times) -- they just kept in the IndexWriter instead of filter before commit in the postRecoveryCallback
These changes first step to fix this.
Then in general we need to enable IndexWriter::Commit in the background before recovery ends.
And then we need to have settings to restrict how many removes could be in the batch.